subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c

```diff
@@ -0,0 +1,647 @@
+		return -EINVAL;
+	}
+
+	char * found = NULL;
+
+	/* replace '-' with "+" */
+	for(found = base64url_string; (found = strchr(found,'-'));) {
+		*found = '+';
+	}
+
+	/* replace '_' with "/" */
+	for(found = base64url_string; (found = strchr(found,'_'));) {
+		*found = '/';
+	}
+
+	/* return number of padding chars required */
+	return (strlen(base64url_string) % 4);
+}
+
+
+static void http_response_cb(struct http_response *rsp,
+			enum http_final_call final_data,
+			void *user_data)
+{
+	struct nrf_cloud_rest_context * rest_ctx = NULL;
+
+	if (user_data) {
+		rest_ctx = (struct nrf_cloud_rest_context *)user_data;
+	}
+
+	if (rest_ctx && rsp->body_found && rsp->body_start) {
+		rest_ctx->response = rsp->body_start;
+	}
+
+	if (final_data == HTTP_DATA_FINAL) {
+		LOG_DBG("HTTP: All data received, status: %u %s",
+			rsp->http_status_code, rsp->http_status);
+
+		if (!rest_ctx) {
+			LOG_WRN("User data not provided...");
+			return;
+		}
+
+		if (rest_ctx) {
+			rest_ctx->status = rsp->http_status_code;
+			rest_ctx->response_len = rsp->content_length;
+		}
+	}
+}
+
+static int generate_auth_header(const char * const tok, char ** auth_hdr_out)
+{
+	if (!tok || !auth_hdr_out)
+	{
+		return -EINVAL;
+	}
+
+	int ret;
+	size_t buff_size = sizeof(AUTH_HDR_BEARER_TEMPLATE) + strlen(tok);
+
+	*auth_hdr_out = k_calloc(buff_size,1);
+	if (!*auth_hdr_out) {
+		return -ENOMEM;
+	}
+	ret = snprintk(*auth_hdr_out, buff_size, AUTH_HDR_BEARER_TEMPLATE, tok);
+	if (ret < 0 || ret >= buff_size) {
+		k_free(*auth_hdr_out);
+		*auth_hdr_out = NULL;
+		return -ENOBUFS;
+	}
+
+	return 0;
+}
+
+int tls_setup(int fd, const char * const tls_hostname)
+{
+	int err;
+	int verify;
+	const sec_tag_t tls_sec_tag[] = {
+		TLS_SEC_TAG,
+	};
+
+	enum {
+		NONE = 0,
+		OPTIONAL = 1,
+		REQUIRED = 2,
+	};
+
+	verify = REQUIRED;
+
+	err = setsockopt(fd, SOL_TLS, TLS_PEER_VERIFY, &verify, sizeof(verify));
+	if (err) {
+		LOG_ERR("Failed to setup peer verification, error: %d", errno);
+		return err;
+	}
+
+	err = setsockopt(fd, SOL_TLS, TLS_SEC_TAG_LIST, tls_sec_tag,
+			 sizeof(tls_sec_tag));
+	if (err) {
+		LOG_ERR("Failed to setup TLS sec tag, error: %d", errno);
+		return err;
+	}
+
+	if (tls_hostname) {
+		err = setsockopt(fd, SOL_TLS, TLS_HOSTNAME, tls_hostname,
+				 strlen(tls_hostname));
+		if (err) {
+			LOG_ERR("Failed to setup TLS hostname, error: %d", errno);
+			return err;
+		}
+	}
+	return 0;
+}
+
+static int socket_timeouts_set(int fd)
+{
+	int err;
+
+	/* Set socket timeouts (send TO also affects TCP connect) */
+	struct timeval timeout = {
+		.tv_sec = 60,
+		.tv_usec = 0
+	};
+
+	err = setsockopt(fd, SOL_SOCKET, SO_SNDTIMEO,
+			 &timeout, sizeof(timeout));
+	if (err) {
+		LOG_ERR("Failed to set socket send timeout, error: %d", errno);
+		return err;
+	}
+
+	err = setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO,
+			 &timeout, sizeof(timeout));
+	if (err) {
+		LOG_ERR("Failed to set socket recv timeout, error: %d/n", errno);
+		return err;
+	}
+
```

Comment:
Nice to see these timeouts being used.  Have you seen them actually time out?

Reply:
I have not